infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -30,96 +38,204 @@ private StateTransfer() {
-                            .prefixTrim(prefixToken);
-                })
-                .forEach(CFUtils::getUninterruptibly);
-
-        if (trimMark > segment.getEnd()) {
-            log.info("stateTransfer: Nothing to transfer, trimMark {} greater than end of segment {}",
-                    trimMark, segment.getEnd());
-            return;
-        }
+            try {
+
+                long trimMark = setTrimOnNewLogUnit(layout, runtime, endpoint);
 
-        // State transfer should start from segment start address or trim mark whichever is lower.
-        long segmentStart = Math.max(trimMark, segment.getStart());
+                if (trimMark > segment.getEnd()) {
+                    log.info("stateTransfer: Nothing to transfer, trimMark {}"
+                                    + "greater than end of segment {}",
+                            trimMark, segment.getEnd());
+                    return true;
+                }
+
+                // State transfer should start from segment start address or trim mark
+                // whichever is higher.
+                final long segmentStart = Math.max(trimMark, segment.getStart());
+                final long segmentEnd = segment.getEnd() - 1;
+                log.info("stateTransfer: Total address range to transfer: [{}-{}] to node {}",
+                        segmentStart, segmentEnd, endpoint);
+
+                // Create batches of STATE_TRANSFER_KNOWN_ADDRESS_BATCH_SIZE addresses.
+                for (long pendingWritesBatch = segmentStart; pendingWritesBatch <= segmentEnd
+                        ; pendingWritesBatch += STATE_TRANSFER_KNOWN_ADDRESS_BATCH_SIZE) {
+
+                    long pendingWritesBatchEnd = Math.min(segmentEnd,
+                            pendingWritesBatch + STATE_TRANSFER_KNOWN_ADDRESS_BATCH_SIZE - 1);
+
+                    // Partition the large batch into chunks to read/write within the RPC timeout.
+                    Iterable<List<Long>> chunks = getMissingEntriesChunks(
+                            layout,
+                            runtime,
+                            chunkSize,
+                            endpoint,
+                            pendingWritesBatch,
+                            pendingWritesBatchEnd);
+
+                    // Read and write in chunks of chunkSize.
+                    for (List<Long> chunk : chunks) {
+                        transferChunk(layout, runtime, endpoint, chunk);
+                    }
+                }
+            } catch (OverwriteException oe) {
 
-        for (long chunkStart = segmentStart; chunkStart < segment.getEnd()
-                ; chunkStart = chunkStart + batchSize) {
-            long chunkEnd = Math.min((chunkStart + batchSize - 1), segment.getEnd() - 1);
+                // If Overwrite Exception is due to Different Data, log the error and abort.
+                if (!oe.getOverWriteCause().equals(OverwriteCause.SAME_DATA)) {
```

Comment:
This doesn't do what it seems like its doing. Look here #1931

Also, even if `OverwriteCause.SAME_DATA` does a byte-wise comparison its not sufficient because two different writes can skip each others writes without even observing the `OverwriteException`. If you really want to be paranoid then multiple writes should be reading each other's writes and verifying them against theirs.

Reply:
Done.